### DISCUSSION TITLE
vector of strings and array_view of string_views conversion conundrum

### DISCUSSION THREAD

**Discussion by positron96** [ETLCPP/etl#543] [Q&A]
This is probably a generic C++ question, but it deals with ETL's hierarchy of classes, so I hope it belongs here.

I want to have a function that accepts a container of string-like objects (vectors of any size of strings of any size and C arrays of C strings). I've created it like this: `func(const etl::array_view<etl::string_view> &arg)`. 

With this I can call it for fixed arrays like this:
```
etl::string_view items[] = { "str1", "str2" };
func(items);
```
Here a `etl::array_view` is created implicitly from a C-array.

Also I can call it for vectors like this:
```
etl::vector<etl::string<L>, N> strings;
/* populate vector here  */
etl::vector<etl::string_view, N> lines;
for(const auto &t: strings) lines.push_back(t);
func(lines);
```
So there is an owning vector `strings` of strings which is copied to a vector `lines` that only stores views of that strings. The function is called for `lines`. I cannot pass a `vector<string>` directly because it's not actually a subclass of `vector<string_view>` (that can be used to create an `array_view<string_view>`).
The question is this: is there a way for me to pass a `vector<string>` directly to my function without creating another `vector<string_view>`? Can I modify function declaration to accept a vector of strings as well as something created from C array of C strings?

**Comment by jwellbelove** [comment]
You could possibly modify `func` to be a template function that takes a const reference to a container that implements `begin()` and `end()`.

```
template <typename TContainer>
func(const TContainer& container)
{
  for (const auto& s : container)
  {
  //...
  }
}
```
